### PR TITLE
Refactor Dockerfile

### LINKED ISSUE #1529: Docker xbuild for arm64 fails on amd64
Building for `linux/arm64` on a `amd64` host fails when using `docker xbuild`.

## Current Behavior

```
$ docker buildx build --platform=linux/arm64 .
[+] Building 0.7s (11/25)
 => [internal] load build definition from Dockerfile                                                                                                                                                       0.0s
 => => transferring dockerfile: 32B                                                                                                                                                                        0.0s
 => [internal] load .dockerignore                                                                                                                                                                          0.0s
 => => transferring context: 35B                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/debian:11-slim                                                                                                                                          0.2s
 => [internal] load metadata for docker.io/lukemathwalker/cargo-chef:latest-rust-1.67.0                                                                                                                    0.0s
 => [chef 1/2] FROM docker.io/lukemathwalker/cargo-chef:latest-rust-1.67.0                                                                                                                                 0.0s
 => [internal] load build context                                                                                                                                                                          0.0s
 => => transferring context: 29.39kB                                                                                                                                                                       0.0s
 => CACHED [stage-3 1/6] FROM docker.io/library/debian:11-slim@sha256:77f46c1cf862290e750e913defffb2828c889d291a93bdd10a7a0597720948fc                                                                     0.0s
 => ERROR [stage-3 2/6] RUN apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*                                                                             0.4s
 => CACHED [chef 2/2] WORKDIR /qdrant                                                                                                                                                                      0.0s
 => CACHED [planner 1/2] COPY . .                                                                                                                                                                          0.0s
 => [planner 2/2] RUN cargo chef prepare --recipe-path recipe.json                                                                                                                                         0.4s
------
 > [stage-3 2/6] RUN apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*:
#0 0.371 exec /bin/sh: exec format error
------
ERROR: failed to solve: executor failed running [/bin/sh -c apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*]: exit code: 1
```

## Steps to Reproduce
1. Must be on a `amd64` host.
2. Invoke `docker xbuild build --platform=linux/arm64 .`
3. Building the docker image fails

## Expected Behavior
Building the image using the above command should succeed.

## Possible Solution
I suspect that one of the stages uses an incorrect architecture, based on the `exec /bin/sh: exec format error` error. Setting this up properly may fix the build issue.

### PR DESCRIPTION
**Pull request by ffuugoo**
Refactored Dockerfile:

- fixed cross-compilation from `aarch64` host to `x86_64` target
  - e.g., fix Docker build on Apple Silicon Macs
  - resolves #1529, closes #1530
- reduce packages/dependencies installed during `builder` stage
- exposed `PROFILE`, `FEATURES`, `RUSTFLAGS` and `LINKER` options/parameters as Dockerfile `ARG`s
- arranged `ARG` directives for better caching

__TODO:__
- [x] check if `lld` is used/works for linkage
  - [x] and figure how to enable it if it's not used, or remove if it doesn't work
  - [x] and/or, utilize [`mold`][mold] linker
    - `mold` is not in the Debian 11 repos...
    - ...but prebuilt Linux binaries are available on GitHub
- [x] document Dockerfile

[mold]: https://github.com/rui314/mold

### CODE REVIEW

**Comment by ffuugoo** [comment]
So, I've tested the Dockerfile in most configurations imaginable: native and cross-comile builds on both ARM and x86 (on M1 Mac, through docker virtualization; takes ages :/), without `lld`, with `lld` _installed_ and with `lld` _enabled for Rust_. However, I've mostly tested _debug and no-LTO release_ builds, as release docker builds with "fat" LTO take literal ages (20+ minutes) on my machine.

My results:
- `lld`, __if present__, speeds up C/C++ dependencies compilation and gives a reliably observable __~20 seconds boost__ (though, it takes 5-10 seconds to be installed)
- __if enabled for Rust__, it seems to make absolutely __no difference__

I've currently pushed a change with `lld` enabled for Rust code, with a simple way to disable it. Could you guys test how does it work on your machines during the day?

- `docker buildx build` would build in __release__ mode, with `lld` __enabled__
- `docker buildx build --build-arg PROFILE=dev` would build in _debug_ mode
- `docker buildx build --build-arg LINKER=` would _disable_ the `lld`
- (and, of course, you use both: `docker buildx build --build-arg PROFILE=dev --build-arg LINKER=`)

___Note the times for `builder 6/8` and `builder 8/8` stages.___

**Comment by ffuugoo** [comment]
Oh, and BTW, is _unstripped debug build for linux __900-1000 MB!?___

**Comment by timvisee** [comment]
> I've currently pushed a change with `lld` enabled for Rust code, with a simple way to disable it. Could you guys test how does it work on your machines during the day?
> 
>     * `docker buildx build` would build in **release** mode, with `lld` **enabled**
> 
>     * `docker buildx build --build-arg LINKER=` would _disable_ the `lld`
> 
>     * `docker buildx build --build-arg PROFILE=dev` would build in _debug_ mode
> 
>     * (and, of course, you use both: `docker buildx build --build-arg PROFILE=dev --build-arg LINKER=`)
> 
> 
> _**Note the times for `builder 6/8` and `builder 8/8` stages.**_

I ran the four commands, see my results below. Notice that I did add `--no-cache`:

| Command | Total time | B6 | B8 |
|:---|---:|---:|---:|
| `docker buildx build --no-cache .` | 357.3s | 90.3s | 240.4s |
| `docker buildx build --no-cache --build-arg LINKER=.` | 351.4s | 88.7s | 239.9s |
| `docker buildx build --no-cache --build-arg PROFILE=dev .` | 171.9s | 107.4s | 32.0s |
| `docker buildx build --no-cache --build-arg LINKER= --build-arg PROFILE=dev .` | 219.8s | 132.4s | 47.4s |

I ran this on Ubuntu (Linux 5.19 x86_64), AMD Ryzen 9 5900X (12/24 cores), 32GB RAM.

I guess that for release builds there are no visible improvements (outside the margin of error).

<details>
<summary>Click here for the logs.</summary>

image

image

image

image


</details>

**Comment by ffuugoo** [comment]
Figured why release Docker builds failed on my Mac... it was RAM. Increased RAM limit available to Docker to 12GB and it works now. Didn't note the build time, though. 🤦‍♀️

**Comment by ffuugoo** [comment]
Rebased to current `dev`. All speedups from #1856 and #1859 present.

Here's a comparison between "cold" builds. Left is this PR (aca7221, [CI link][refactor-cold-build]), right is #1859 (d0933e5, [CI link][1859-cold-build]).
<details>
<summary>Screenshots</summary>
Screenshot 2023-05-09 at 10 15 23
</details>

And here's a comparison between "cold" build (aca7221, [CI link][refactor-cold-build-summary]) and "hot" build (dbbac23, [CI link][refactor-hot-build-summary]) in this PR.
<details>
<summary>Screenshots</summary>
Screenshot 2023-05-09 at 10 40 19 Screenshot 2023-05-09 at 10 40 24
</details>

And, for the reference, "cold" build (d0933e5, [CI link][1859-cold-build-summary]) and "hot" build (d4e4453, [CI link][1859-hot-build-summary]) in the #1859.

<details>
<summary>Screenshots</summary>
Screenshot 2023-05-09 at 10 55 28 Screenshot 2023-05-09 at 10 55 32
</details>

[refactor-cold-build]: https://github.com/qdrant/qdrant/actions/runs/4923671415/jobs/8796327074
[1859-cold-build]: https://github.com/qdrant/qdrant/actions/runs/4917829435/jobs/8783381655
[refactor-cold-build-summary]: https://github.com/qdrant/qdrant/actions/runs/4923671415
[refactor-hot-build-summary]: https://github.com/qdrant/qdrant/actions/runs/4924031125
[1859-cold-build-summary]: https://github.com/qdrant/qdrant/actions/runs/4917829435
[1859-hot-build-summary]: https://github.com/qdrant/qdrant/actions/runs/4918068509

**Comment by ffuugoo** [comment]
Rebased on recent `dev`.

**Comment by generall** [comment]
Here is my checklist:

- [ ] docker buildx works for both - amd and arm
- [ ] regular `docker build .` works
- [ ]  Docker uses release profile by default, but debug provile configured in CI
- [ ] Caches work
- [ ] mold is used as much as possible, but at least on amd arch
- [ ] changes in code does not invalidate dependencies cache

If those checked, then - LGTM

**Comment by ffuugoo** [comment]
@generall I'll verify all of these once again later, but unless something got accidentally broken recently, __all__ of them should be good.

**Comment by timvisee** [comment]
Trying to build for arm64 (on amd64) fails for me with this command:

```bash
$ docker buildx build --platform='linux/arm64' .
...
 => ERROR [qdrant 2/7] RUN apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*                                                         1.0s
 => [planner 2/3] COPY . .                                                                                                                                                            1.0s
------
 > [qdrant 2/7] RUN apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*:
#0 0.642 exec /bin/sh: exec format error
------
ERROR: failed to solve: executor failed running [/bin/sh -c apt-get update     && apt-get install -y ca-certificates tzdata     && rm -rf /var/lib/apt/lists/*]: exit code: 1
```

**Comment by ffuugoo** [comment]
@generall I've tested the Dockerfile once again and marked most items in the checklist. I need help testing on **AMD64 host** (as I only have macOS, which is ARM).

I think sufficient tests would be:
- `docker build .`
- `docker buildx build .`
  - then modify any Rust file, re-run `docker buildx build .` and check that steps **2 to 7** of the `builder` stage are cached (look for the `CACHED [builder 2/9]` lines)
  - (it's kinda optional, but it's a quick test)
- `docker buildx build . --platform linux/arm64`
- (also, you better run all these steps with `--build-arg PROFILE=ci` to not wait for eternity for LTO to complete)

**Comment by timvisee** [comment]
@ffuugoo I'll give these a spin. I'm on Linux with amd64.

**Comment by timvisee** [comment]
@ffuugoo All builds (`build`, `buildx`, `buildx for arm64`) work fine on my machine!

For cross compilation with `buildx` I did have to explicitly set a custom builder like `--builder=container`, which I created with `docker buildx create --driver=docker-container --name=container`. This has to do with my setup though, and don't think it can be improved upon in this PR.

Cache works with `buildx`.

I used `--build-arg PROFILE=ci` on all.

File: Dockerfile [hunk 1]
```diff
@@ -1,70 +1,101 @@
-# Leveraging the pre-built Docker images with
-# cargo-chef and the Rust toolchain
+# Cross-compiling using Docker multi-platform builds/images and `xx`.
+#
+# https://docs.docker.com/build/building/multi-platform/
+# https://github.com/tonistiigi/xx
+FROM --platform=${BUILDPLATFORM:-linux/amd64} tonistiigi/xx AS xx
+
+# Utilizing Docker layer caching with `cargo-chef.
```
Comment (timvisee) [inline Dockerfile:7]:
```suggestion
# Utilizing Docker layer caching with `cargo-chef`.
```

**Review by timvisee — APPROVED** [review]
Great work on this! Don't see any real problems based on a quick inspection.

File: Dockerfile [hunk 2]
```diff
@@ -1,70 +1,101 @@
...
-ARG RUST_BUILD_PROFILE=release
-
-ARG MOLD_VERSION=1.11.0
+FROM --platform=${BUILDPLATFORM:-linux/amd64} lukemathwalker/cargo-chef:latest-rust-1.68.1 AS chef
 
-# # Choose MOLD arch based on TARGETARCH: amd64 -> x86_64, arm64 -> aarch64
 
 FROM chef AS planner
+WORKDIR /qdrant
 COPY . .
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply Dockerfile:15]: Everything, except for paths in `.dockerignore` file.

[the remark this answers is not in the document]
Comment (timvisee) [reply Dockerfile:15]: Awesome, wasn't aware of that! So it [excludes](https://github.com/qdrant/qdrant/blob/60d226385ddc68101d3e41ab450f8e50e3191681/.dockerignore#L2) `target/`. Fantastic!

**Pull request merged**